File: 백준/Gold/1738. 골목길/골목길.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException{
    	BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    	st = new StringTokenizer(br.readLine());
    	N = Integer.parseInt(st.nextToken());
    	M = Integer.parseInt(st.nextToken());
    	dist = new long[N+1];
    	
    	edges = new Edge[M];
    	Arrays.fill(dist, INF);
    	for(int i = 0; i < M; i++) {
    		st = new StringTokenizer(br.readLine());
    		int from = Integer.parseInt(st.nextToken());
    		int to = Integer.parseInt(st.nextToken());
    		int cost = Integer.parseInt(st.nextToken());
    		
    		edges[i] = new Edge(from, to, cost);
    	}
    	
    	prev = new int[N+1];
    	dist[1] = 0;
    	for(int i = 1; i<=N; i++) {
    		for(Edge e : edges) {
    			if(dist[e.from] == INF)
    				continue;
    			
    			if(dist[e.to] < dist[e.from] + e.cost) {
    				prev[e.to] = e.from;
    				dist[e.to] = dist[e.from] + e.cost;
    			}
    		}
    	}
    	
    	for(Edge e : edges) {
    		if(dist[e.from] == INF || dist[e.to] >= dist[e.from ] + e.cost)
				continue;
    		
    		if(canReach(1,e.from) && canReach(e.from, N)) {
    			System.out.println(-1);
    			return;
    		}
			
    	}
    	Stack<Integer> stk = new Stack<>();
    	int cur = N;
    	while(cur != 0) {
    		stk.push(cur);
    		cur = prev[cur];
    	}
    	
    	StringBuilder sb = new StringBuilder();
    	while(!stk.isEmpty()) {
    		sb.append(stk.pop()).append(' ');
    	}
    	
    	System.out.println(sb);
    	
    } 
    
    static boolean canReach(int start, int end) {
    	boolean visited[] = new boolean[N+1];
    	Queue<Integer> q = new ArrayDeque<>();
    	q.add(start);
    	
    	visited[start] = true;
    	
    	while(!q.isEmpty()) {
    		int cur = q.poll();
    		
    		if(cur == end)
    			return true;
    		
    		for(Edge e : edges) {
    			if(e.from != cur || visited[e.to])
    				continue;
    			q.add(e.to);
    			visited[e.to] = true;
    		}
    	}
    	
    	return false;
    }
    
    
    static int N,M;
    static StringTokenizer st;
    static long dist[];
    static int prev[];
    static Edge edges[];
    static long INF = Long.MIN_VALUE;
    static class Edge{
    	int from,to,cost;
    	public Edge(int from, int to, int cost) {
    		this.from = from;
    		this.to = to;
    		this.cost = cost;
    	}
    }
}

File: 백준/Gold/1738. 골목길/골목길.py
from collections import namedtuple, deque

from pprint import pprint

def canUpdate(e):
    return dist[e.v] >= dist[e.u] + e.cost

def canReach(start, end):
    q = deque([start])
    visited = [False] * (N+1)
    visited[start] = True

    while q:
        cur = q.popleft()

        if cur == end:
            return True

        for e in edges:
            if e.u == cur and not visited[e.v]:
                visited[e.v] = True
                q.append(e.v)

    return False

if __name__ == '__main__':
    N,M = map(int,input().split())
    Edge = namedtuple('Edge', ['u', 'v', 'cost'])
    edges = []
    INF = -float('inf')
    for i in range(M):
        _from, _to, _cost = map(int,input().split())
        edges.append(Edge(_from,_to,_cost))


    dist = [INF] * (N+1)

    dist[1] = 0
    prev = [0] * (N+1)
    for i in range(N):
        for edge in edges:
            if dist[edge.u] == INF:
                continue

            if dist[edge.v] < dist[edge.u] + edge.cost:
                dist[edge.v] = dist[edge.u] + edge.cost
                prev[edge.v] = edge.u

    # pprint(dist)
    # pprint(prev)
    for edge in edges:
        if dist[edge.u] == INF or canUpdate(edge):
            continue

        if canReach(1, edge.u) and canReach(edge.u, N):
            # print(edge)
            print(-1)
            exit()


    ans = []
    cur = N
    while cur != 0:
        ans.append(cur)
        cur = prev[cur]
    print(' '.join(map(str, ans[::-1])))